docs/design-proposals/2021-11-24-graceful-reschedule-tikv-pod.md

```diff
@@ -0,0 +1,107 @@
+# gracefully reschedule tikv pod
+
+TiDB-Operator issue: [#4215](https://github.com/pingcap/tidb-operator/issues/4215)
+
+## Summary
+
+Support gracefully reschedule a tikv pod. The implemtation is necessarrily the combination of three operations:
+
+1. Add a PD evict-leader-scheduler to transfer the leader.
+2. When leader count drop down to zero, delete the pod to let it re-create and reschedule.
+3. Remove the evict-leader-scheduler.
+
+## Motivation
+
+### Goals
+
+- Provide a way for user to gracefully reschedule a tikv pod.
+- The pod must be gracefully stop, to be specify, evict leaders before delete the pod and must work nomal when recreate the pod.
+
+### Non-Goals
+
+- Provider a way to make tikv graceful stop when just recive a `SIGTERM` signal.
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+suppost `tikv-0` is running at node `node0`,  there are many free resource at other nodes, I would like to make tikv-0 run at other node and shutdown `node0` to reduce the cost. I may apply the flowing operations:
+
+1. run `kubectl cordon node0` to mark `node0`as unschedulable.
+2. delete pod `tikv-0` to let it re-create and reschedule to other nodes.
+3. draine `node-0` in the nomal way as [safely drain a node](https://kubernetes.io/docs/tasks/administer-cluster/safely-drain-node/) and shotdown `node0`
+
+### Risks and Mitigations
+
+- PD/TiKV fails to evict all leader and we may just keep checking it.
+
+## Design Details
+
+Support user to add an annotation to CRD `TidbCluster` to trigger an gracefule reschedule.
+
+annotation key: `tikv.tidb.pingcap.com/reschedule`
+
+annotation value is JSON in the flowing struct:
+
+```go
+{
+  // the flowing field set by user.
+  NodeName *string
```

Comment:
Why not support specify more than one node?

Reply:
Why nodeName is needed?

Reply:
generally, we just want to take one node per time?
you can do it one by one.

Reply:
> generally, we just want to take one node per time?

but if the user set more than one node, operator can still do it one by one, right?